File: immiglensBE/app/schemas/audit_log.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class AuditLogOut(BaseModel):
    id: int

    # Actor
    actor_id: Optional[int] = None
    actor_type: str = "user"
    actor_email: Optional[str] = None
    actor_name: Optional[str] = None

    # Action
    action: str
    status: str = "success"

    # Entity
    entity_type: str
    entity_id: Optional[str] = None
    entity_label: Optional[str] = None

    # Scope
    employer_id: Optional[int] = None
    position_id: Optional[int] = None

    # Payload
    description: Optional[str] = None
    old_data: Optional[dict[str, Any]] = None
    new_data: Optional[dict[str, Any]] = None
    # The ORM column is named metadata_ to avoid collision with SQLAlchemy's
    # reserved MetaData. validation_alias maps it back to the API field name.
    metadata: Optional[dict[str, Any]] = Field(None, validation_alias="metadata_")

    # Context
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    source: str = "api"

    created_at: datetime

    model_config = {"from_attributes": True, "populate_by_name": True}

File: immiglensFE/src/types/audit_log.ts
export interface AuditLog {
  id: number
  // New fields (v2)
  actor_id: number | null
  actor_type: string | null
  actor_email: string | null
  actor_name: string | null
  entity_type: string
  entity_id: string | null
  entity_label: string | null
  status: string | null
  description: string | null
  metadata: Record<string, unknown> | null
  user_agent: string | null
  source: string | null
  // Retained fields
  action: string
  employer_id: number | null
  position_id: number | null
  old_data: Record<string, unknown> | null
  new_data: Record<string, unknown> | null
  ip_address: string | null
  created_at: string
  // Deprecated — kept for backward compat during migration
  /** @deprecated use actor_id */
  user_id?: number | null
  /** @deprecated use actor_email */
  user_email?: string | null
  /** @deprecated use actor_name */
  user_name?: string | null
  /** @deprecated use entity_type */
  resource_type?: string
  /** @deprecated use entity_id */
  resource_id?: number | null
}
